python/paddle/distributed/passes/auto_parallel_sharding.py

```diff
@@ -0,0 +1,694 @@
+
+    return dp_group, sharding_group
+
+
+def _is_gradient_clip_op(op):
+    return op.desc.has_attr("op_namescope") \
+        and op.desc.attr("op_namescope").startswith("/gradient_clip")
+
+
+def _is_weight_decay_op(op):
+    return op.desc.has_attr("op_namescope") \
+        and op.desc.attr("op_namescope").startswith("/regularization")
+
+
+def _is_param_grad_fp32_cast_op(block, op):
+    if not is_backward_op(op):
+        return False
+    if not _is_desired_cast_op(block, op, core.VarDesc.VarType.FP16,
+                               core.VarDesc.VarType.FP32):
+        return False
+    output_name = op.desc.output_arg_names()[0]
+    base_name = output_name[:output_name.find("@")]
+    if not block.has_var(base_name):
+        return False
+    return block.var(base_name).is_parameter
+
+
+def _is_param_fp16_cast_op(block, op, params):
+
+    if is_optimizer_op(op):
+        return False
+    if not _is_desired_cast_op(block, op):
+        return False
+    input_name = op.desc.input_arg_names()[0]
+    if input_name not in params:
+        return False
+    return True
+
+
+def _is_desired_cast_op(block,
+                        op,
+                        src_var_type=core.VarDesc.VarType.FP32,
+                        dst_var_type=core.VarDesc.VarType.FP16):
+    if op.type != "cast":
+        return False
+    assert (len(op.desc.input_arg_names()) == 1)
+    assert (len(op.desc.output_arg_names()) == 1)
+    input_var = block.var(op.desc.input_arg_names()[0])
+    output_var = block.var(op.desc.output_arg_names()[0])
+
+    if input_var.dtype != src_var_type or \
+        output_var.dtype != dst_var_type:
+        return False
+
+    return True
+
+
+def _get_base_name_from_grad_name(grad_name):
+    base_name = None
+    if ".cast_fp16@GRAD" in grad_name:
+        base_name = grad_name[:grad_name.find(".cast_fp16@GRAD")]
+    elif "@GRAD" in grad_name:
+        base_name = grad_name[:grad_name.find("@GRAD")]
+    return base_name
+
+
+def _is_param_grad_allreduce_op(op, block, dp_ring_ids):
+
+    if not is_backward_op(op):
+        return False
+    if op.type != "c_allreduce_sum":
+        return False
+    if op.attr('ring_id') not in dp_ring_ids:
+        return False
+
+    output_name = op.output_arg_names[0]
+    base_name = _get_base_name_from_grad_name(output_name)
+
+    if not block.has_var(base_name):
+        return False
+
+    return block.var(base_name).is_parameter
+
+
+def _inference_data_parallel_group_for_operator(rank_id, op, dist_context):
+
+    dp_group = None
+    for input_name in op.input_arg_names:
+        if not is_parameter_related(input_name, op.block):
+            dist_attr = dist_context.get_op_dist_attr_for_program(op)
+            process_mesh = dist_attr.process_mesh
+            input_dim_mapping = dist_attr.get_input_dims_mapping(input_name)
+            mesh_shape = process_mesh.topology
+            # TODO replace with specific batch size dimension
```

Comment:
I suggest that each `TODO` comment should add **who is responsible TODO**.

Reply:
fixed~